### PR TITLE
feat(types): NavigationGuardNext

### PR DESCRIPTION
**Pull request by beary**
Expose a `Next` type for this usage:
```typescript
// ...
import { Route, Next } from 'vue-router'

@Component
export default class Verify extends Vue {
  // ...
  private beforeRouteEnter(to: Route, from: Route, next: Next) {
    next(vm => {
      vm.$data.username = decodeURIComponent(to.query.username || '')
    })
  }
}
```

### CODE REVIEW

**Comment by kevlarr** [comment]
Exactly the PR I was about to submit.. @posva How do PRs get reviewed/merged?

**Comment by posva** [comment]
I will come at it when I have time, after other Bugs and features

**Comment by voool** [comment]
As a workaround, to grab the `next` interface with Typescript you can do
```
import { NavigationGuard } from 'vue-router';
type Next = Parameters<NavigationGuard>[2];
```

**Comment by kevlarr** [comment]
Nice @voool, that's really clean! I didn't know that was possible - I've been copying the type def into my own code, which is more verbose and also more brittle (in case the def changes at all).

**Comment by andredewaard** [comment]
Any progress on when this will be merged?

**Comment by faustbrian** [comment]
@posva any progress on this?

**Comment by d2phap** [comment]
Any update?

File: types/router.d.ts [hunk 1]
```diff
@@ -7,10 +7,11 @@ type ErrorHandler = (err: Error) => void;
 export type RouterMode = "hash" | "history" | "abstract";
 export type RawLocation = string | Location;
 export type RedirectOption = RawLocation | ((to: Route) => RawLocation);
+export type Next<V extends Vue = Vue> = (to?: RawLocation | false | ((vm: V) => any) | void) => void;
```
Comment (posva) [inline types/router.d.ts]:
```suggestion
export type NavigationGuardNext<V extends Vue = Vue> = (to?: RawLocation | false | ((vm: V) => any) | void) => void;
```

File: types/index.d.ts [hunk 2]
```diff
@@ -12,5 +12,6 @@ export {
   RouteRecord,
   Location,
   Route,
-  NavigationGuard
+  NavigationGuard,
+  Next
```
Comment (posva) [inline types/index.d.ts]:
```suggestion
  NavigationGuardNext
```

File: types/router.d.ts [hunk 3]
```diff
@@ -7,10 +7,11 @@ type ErrorHandler = (err: Error) => void;
...
 export type NavigationGuard<V extends Vue = Vue> = (
   to: Route,
   from: Route,
-  next: (to?: RawLocation | false | ((vm: V) => any) | void) => void
+  next: Next<V>
```
Comment (posva) [inline types/router.d.ts]:
```suggestion
  next: NavigationGuardNext<V>
```

**Review by posva — CHANGES_REQUESTED** [review]
I'm renaming the type to match vue-router-next
One thing that is problematic is that only `beforeRouteEnter`'s `next` allows a callback. So I think we should export a type without the callback and make a specific type for `beforeRouteEnter` where we change `next` to include `(vm: V) => any`

**Review by beary — COMMENTED** [review]
Looks good :smiley:

**Review by beary — COMMENTED** [review]
:+1:

**Comment by beary** [comment]
I am not familiar with the operation of pr, if it interferes with people, I am sorry.
Or if the commit record is too confusing, you can close this pull request and create a new one.

**Comment by posva** [comment]
No worries for the commit history @beary, we squash commits at the end! Let me know if this is things are not clear and I will finish this up

**Comment by beary** [comment]
@posva 
I read your changes, I think it is very good, And there is nothing I need to add!

**Comment by posva** [comment]
@beary There is a change left: https://github.com/vuejs/vue-router/pull/2497#pullrequestreview-408928568

**Review by beary — COMMENTED** [review]
Reviewed

**Review by beary — COMMENTED** [review]
I submitted it for review, but it is still in a reviewable state

**Review by posva — CHANGES_REQUESTED** [review]
One thing that is problematic is that only `beforeRouteEnter`'s `next` allows a callback. So I think we should export a type without the callback and make a specific type for `beforeRouteEnter` where we change `next` to include `(vm: V) => any`

**Comment by posva** [comment]
@beary because the review is the same

**Review by beary — COMMENTED** [review]
Reviewed

**Comment by posva** [comment]
what? see https://github.com/vuejs/vue-router/pull/2497#pullrequestreview-408976579
if you don't see where to add, that's fine, just let me know and I will pick it but let's not go in circles and waste each other time

**Comment by beary** [comment]
@posva 
Sorry I couldn't add it recently

**Pull request merged**